File: 백준/Gold/17140. 이차원 배열과 연산/이차원 배열과 연산.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static int r, c, k;
    static int[][] map;


    public static void main(String[] args) throws IOException {

        input();

        int rowLen = 3;
        int colLen = 3;

        int time = 0;

        Map<Integer, Integer> cntMap = new HashMap<>();
        while(time<100){

            if(map[r][c]==k) {
                break;
            }

            if(rowLen>=colLen){

                //map에 넣기

                int max= 0;
                for(int i=0; i<rowLen; i++){
                    cntMap.clear();
                    for(int j=0; j<colLen; j++){
                        if(map[i][j]!=0) cntMap.put(map[i][j], cntMap.getOrDefault(map[i][j],0)+1);
                    }

                    //map을 value 순으로 정렬

                    List<Integer> list = sort(cntMap);


                    int idx =0;
                    Arrays.fill(map[i], 0);
                    for(int j=0; j<Math.min(100, list.size()); j++){
                        map[i][idx++] = list.get(j);
                        map[i][idx] = cntMap.get(map[i][idx-1]);
                        idx++;
                    }

                    max = Math.max(Math.min(100, list.size()*2), max);
               }

                colLen = max;

            }else{

                int max =0;
                for(int c =0; c<colLen; c++){
                    cntMap.clear();
                    for(int r=0; r<rowLen; r++){
//                        if(map[r][c]==0) break;
                        if(map[r][c]!=0) cntMap.put(map[r][c], cntMap.getOrDefault(map[r][c],0)+1);
                    }

                    //map을 value 순으로 정렬

                    List<Integer> list = sort(cntMap);


                    int idx =0;

                    for(int j=0; j<100; j++){
                        map[j][c] = 0;
                    }
                    for(int j=0; j<Math.min(100, list.size()); j++){
                        map[idx++][c] = list.get(j);
                        map[idx][c] = cntMap.get(map[idx-1][c]);
                        idx++;
                    }

                    max = Math.max(Math.min(list.size()*2, 100), max);
                }

                rowLen = max;

            }

            time++;


        }

        System.out.println(map[r][c]==k ? time : -1);



    }

    static List<Integer> sort(Map<Integer, Integer> cntMap){
        //map을 value 순으로 정렬

        List<Integer> list = new ArrayList<>(cntMap.keySet());
        list.sort(new Comparator<Integer>() {
            @Override
            public int compare(Integer o1, Integer o2) {
                if(cntMap.get(o1)-cntMap.get(o2)==0){
                    return o1-o2;
                }
                return cntMap.get(o1)-cntMap.get(o2);
            }
        });

        return list;
    }

    static void input() throws IOException {
        st = new StringTokenizer(br.readLine());

        r = Integer.parseInt(st.nextToken())-1;
        c = Integer.parseInt(st.nextToken())-1;
        k = Integer.parseInt(st.nextToken());

        map = new int[100][100];


        for(int i=0; i<3; i++){
            st = new StringTokenizer(br.readLine());
            for(int j=0; j<3; j++){
                map[i][j] = Integer.parseInt(st.nextToken());
            }
        }
    }
}

File: 백준/Gold/17140. 이차원 배열과 연산/이차원 배열과 연산.py
"""
행 연산은 좀 쉬울듯 ? 열 연산 어쩌지 배열 회전할까


일단 행 연산부터 생각하자.
그 행에 있는 값을 key로 개수를 value로 딕셔너리로 만들기
 그 후 items로 k, v를 튜플로 lst에 넣고 정렬해서 (정렬할 땐 value우선)
 해당 행에 끼워넣기 !

 배열 크기 맞추기 max len을 찾아서 그만큼 0으로 채워준다



연산하고 행 / 열 길이 보고 행연산 열연산 결정. .!
arr[r][c]의 값이 k 가 되면 종료 혹은 time이 101 되면 종료

"""
R, C, K = map(int, input().split())

arr = [list(map(int, input().split())) for _ in range(3)]
time = 0
while 1:
    row_len = len(arr)
    col_len = len(arr[0])
    if time >100 or ( R-1 < row_len and C-1<col_len and arr[R-1][C-1] == K):
        break

    #행 / 열 크기 확인
    #배열 transpose
    if row_len < col_len:
        arr = list(map(list, zip(*arr)))

    #연산하자
    for i in range(max(row_len, col_len)):
        tmp_dict = {}
        for j in range(min(row_len, col_len)):
            if arr[i][j] == 0:
                continue
            tmp_dict[arr[i][j]] = tmp_dict.get(arr[i][j],0)+1

        lst = []
        for k, v in tmp_dict.items():
            lst.append([k,v])
        lst.sort(key=lambda x:(x[1], x[0]))
        arr[i] = []
        for a in lst:
            arr[i].extend(a)
        arr[i] = arr[i][:100]
    # 0채워주자
    mx = min(row_len, col_len)
    for i in range(len(arr)):
        mx = max(mx, len(arr[i]))
    for i in range(len(arr)):
        while len(arr[i]) < mx:
            arr[i].append(0)



    #transpose했었다면 되돌리자
    if row_len<col_len:
        # for i in range(len(arr)):
        #     print(arr[i])
        # print()
        arr = list(map(list, zip(*arr)))

    time += 1
    # for i in range(len(arr)):
    #     print(arr[i])
    # print()


print(time if time<=100 else -1)